other/cmr-exchange/authz/src/cmr/authz/token.clj

```diff
@@ -58,27 +71,73 @@
       (find-xml [:token_info :user_name])
       first))
 
+(defn parse-lp-username
+  [body]
+  (:uid (:body (json/parse-string body true))))
+
+(defn add-basic-auth
+  ([user pass]
+   (add-basic-auth {} user pass))
+  ([req user pass]
+   (assoc req :basic-auth [user pass])))
+
+(defn get-client-token
+  "Get the CMR client bearer token from from the cache."
+  [system]
+  (let [base-url (config/get-edl-rest-url system)
+        url (str base-url client-token-resource)]
+    (request/async-post
+     url
+     (-> {}
+         (request/add-accept "application/json")
+         (request/add-basic-auth (config/get-edl-username system) (config/get-edl-password system)))
+     {}
+     #(response/general-response-handler % response/error-handler parse-bearer-token))))
+
+(defn verify-edl-token-locally
+  "Uses the EDL public key to verify jwt tokens locally."
+  [system token]
+  (try
+    (log/trace "edl public key= " (config/get-jwt-public-key system))
+    (let [public-key (buddy-keys/jwk->public-key (json/parse-string (config/get-jwt-public-key system) true))
+          bearer-stripped-token (string/replace token #"Bearer\W+" "")
+          decrypted-token (jwt/unsign bearer-stripped-token public-key {:alg :rs256})]
+      (:uid decrypted-token))
+    (catch Exception ex
+      (log/error (.getMessage ex))
+      (log/debug "JWT local token verification failed, trying launchpad instead."))))
+
 (defn get-token-info
   "Query the CMR Access Control API for information assocated with the given
   token."
-  [base-url token]
-  (let [url (str base-url token-info-resource)
-        data (str "id=" token)]
-    (log/trace "Making token-info query to ECHO REST:" url)
-    (request/async-post
-      url
-      (-> {:body data}
-          (http/add-options)
-          (request/add-token-header token)
-          (request/add-form-ct))
-      {}
-      #(response/body-only-response-handler % response/error-handler parse-username))))
+  [system token]
+  (let [base-url (config/get-edl-rest-url system)
+        url (str base-url token-info-resource)]
+    (log/trace "Decoding jwt token to get token info")
+    (if-let [token-info (verify-edl-token-locally system token)]
+      (do
+        (log/trace "user info found for jwt token, token-info=" token-info)
+        (deliver (promise) token-info))
+      (let [bearer-token @(get-client-token system)]
+        (if (:errors bearer-token)
+          (deliver (promise) bearer-token)
+          (do
+            (log/trace "retrieved cmr client token= " bearer-token)
+            (log/trace "user info not found for jwt token, attempting to authenticate user via launchpad")
+            (request/async-post
+             url
+             (-> {:body (str "token=" token)}
+                 (http/add-options)
+                 (request/add-token-header (str "Bearer " (:body bearer-token))))
```

Comment:
Will this ever possibly receive launchpad tokens that don't use `Bearer `?

Reply:
That is the cmr client token, so no.